src/alexandria3k/data_sources/pubmed.py

```diff
@@ -0,0 +1,316 @@
+        return (self.files_cursor.Rowid() << 14) | (self.element_index)
+
+    # pylint: disable=arguments-differ
+    def Filter(self, index_number, _index_name, constraint_args):
+        """Always called first to initialize an iteration to the first row
+        of the table according to the index"""
+        self.files_cursor.Filter(index_number, _index_name, constraint_args)
+        self.eof = self.files_cursor.Eof()
+        if index_number & ROWID_INDEX:
+            # This has never happened, so this is untested
+            self.element_index = constraint_args[1]
+        else:
+            self.element_index = 0
+
+    def Next(self):
+        """Advance to the next item."""
+        self.element_index += 1
+        if self.element_index >= len(self.files_cursor.items):
+            self.element_index = 0
+            self.files_cursor.Next()
+            self.eof = self.files_cursor.eof
+
+    def current_row_value(self):
+        """Return the current row. Not part of the apsw API."""
+        return self.files_cursor.current_row_value()[self.element_index]
+
+    def Close(self):
+        """Cursor's destructor, used for cleanup"""
+        self.files_cursor.Close()
+
+    def container_id(self):
+        """Return the id of the container containing the data being fetched.
+        Not part of the apsw API."""
+        return self.files_cursor.Rowid()
+
+
+class VTSource:
+    """Virtual table data source. This gets registered with the apsw
+    Connection through createmodule in order to instantiate the virtual
+    tables."""
+
+    def __init__(self, data_directory, sample):
+        self.data_files = DataFiles(data_directory, sample, ".gz")
+        self.table_dict = {t.get_name(): t for t in tables}
+        self.sample = sample
+
+    def Create(self, _db, _module_name, _db_name, table_name):
+        """Create a virtual table with the specified name"""
+        table = self.table_dict[table_name]
+
+        return table.table_schema(), StreamingCachedContainerTable(
+            table,
+            self.table_dict,
+            self.data_files.get_file_array(),
+            self.sample,
+        )
+
+    Connect = Create
+
+    def get_container_iterator(self):
+        """Return an iterator over the data files identifiers"""
+        return self.data_files.get_container_iterator()
+
+    def get_container_name(self, fid):
+        """Return the name of the file corresponding to the specified fid"""
+        return self.data_files.get_container_name(fid)
+
+
+# https://www.nlm.nih.gov/bsd/mms/medlineelements.html lists all abbreviations
+tables = [
+    TableMeta(
+        "pubmed_articles",
+        columns=[
+            ColumnMeta("id", getter("MedlineCitation/PMID")),
+            ColumnMeta("container_id"),
+            ColumnMeta(
+                "doi",
+                getter_by_attribute(
+                    "IdType", "doi", "PubmedData/ArticleIdList/ArticleId"
+                ),
```

Comment:
Normalize this to lowercase.

Reply:
This is part of the source XML, so I don't think this can be lowercase 
<img width="288" alt="image" src="https://github.com/dspinellis/alexandria3k/assets/33267625/deb29a93-0473-42c0-abbf-a05c3d578e7e">

Reply:
Sorry I wasn't clear enough. I am suggesting that you normalize the obtained DOI into lowercase to match the normalization done in other parts of a3k.

Reply:
Aaah I see, I'll look at the implementation for CrossRef and see what can be reused

Reply:
The doi has been changed to lowercase